From 8a4c806aa8770700db2ab8898d5eb36ceb3491bf Mon Sep 17 00:00:00 2001
From: Jeff Moyer <jmoyer@redhat.com>
Date: Thu, 29 May 2014 21:11:21 -0400
Subject: [block] mtip32xx: Unmap the DMA segments before completing the IO request

Message-id: <1401397882-5166-2-git-send-email-jmoyer@redhat.com>
Patchwork-id: 81191
O-Subject: [RHEL7 PATCH 1/2] [block] mtip32xx: Unmap the DMA segments before completing the IO request
Bugzilla: 1102281
RH-Acked-by: David Milburn <dmilburn@redhat.com>

This is a backport of the following commit:

  commit 368c89d7ac70f937c93cd6f3b65bcfdfb3ba794f
  Author: Felipe Franciosi <felipe@paradoxo.org>
  Date:   Thu Mar 13 14:34:21 2014 +0000

    If the buffers are unmapped after completing a request, then stale data
    might be in the request.

    Signed-off-by: Felipe Franciosi <felipe@paradoxo.org>
    Cc: stable@kernel.org
    Signed-off-by: Jens Axboe <axboe@fb.com>

This patch is required for the next one to apply cleanly, but is also
important enough to include in z-stream.  Micron has agreed to test
the fix, as we don't have the hardware in our lab.

brew: http://brewweb.devel.redhat.com/brew/taskinfo?taskID=7512189
bz: 1102281

Signed-off-by: Jeff Moyer <jmoyer@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/block/mtip32xx/mtip32xx.c b/drivers/block/mtip32xx/mtip32xx.c
index 3ea4e13..11831bf 100644
--- a/drivers/block/mtip32xx/mtip32xx.c
+++ b/drivers/block/mtip32xx/mtip32xx.c
@@ -245,6 +245,12 @@ static void mtip_async_complete(struct mtip_port *port,
    "Command tag %d failed due to TFE\n", tag);
  }
 
+ /* Unmap the DMA scatter list entries */
+ dma_unmap_sg(&dd->pdev->dev,
+  command->sg,
+  command->scatter_ents,
+  command->direction);
+
  /* Upper layer callback */
  if (likely(command->async_callback))
   command->async_callback(command->async_data, cb_status);
@@ -252,12 +258,6 @@ static void mtip_async_complete(struct mtip_port *port,
  command->async_callback = NULL;
  command->comp_func = NULL;
 
- /* Unmap the DMA scatter list entries */
- dma_unmap_sg(&dd->pdev->dev,
-  command->sg,
-  command->scatter_ents,
-  command->direction);
-
  /* Clear the allocated and active bits for the command */
  atomic_set(&port->commands[tag].active, 0);
  release_slot(port, tag);
@@ -688,6 +688,12 @@ static void mtip_timeout_function(unsigned long int data)
     */
    writel(1 << bit, port->completed[group]);
 
+   /* Unmap the DMA scatter list entries */
+   dma_unmap_sg(&port->dd->pdev->dev,
+     command->sg,
+     command->scatter_ents,
+     command->direction);
+
    /* Call the async completion callback. */
    if (likely(command->async_callback))
     command->async_callback(command->async_data,
@@ -695,12 +701,6 @@ static void mtip_timeout_function(unsigned long int data)
    command->async_callback = NULL;
    command->comp_func = NULL;
 
-   /* Unmap the DMA scatter list entries */
-   dma_unmap_sg(&port->dd->pdev->dev,
-     command->sg,
-     command->scatter_ents,
-     command->direction);
-
    /*
     * Clear the allocated bit and active tag for the
     * command.
-- 
1.7.1